File: leetcode-algorithms/155. Min Stac/Solution.cpp
// Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.
//
// push(x) -- Push element x onto stack.
// pop() -- Removes the element on top of the stack.
// top() -- Get the top element.
// getMin() -- Retrieve the minimum element in the stack.
// Example:
// MinStack minStack = new MinStack();
// minStack.push(-2);
// minStack.push(0);
// minStack.push(-3);
// minStack.getMin();   --> Returns -3.
// minStack.pop();
// minStack.top();      --> Returns 0.
// minStack.getMin();   --> Returns -2.
class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() {

    }

    void push(int x) {
        vstack.push(x);
        if (minstack.empty()) minstack.push(x);
        else minstack.push(min(minstack.top(), x));
    }

    void pop() {
        if (!vstack.empty()) {
            vstack.pop();
            minstack.pop();
        }
    }

    int top() {
        if (!vstack.empty()) {
            return vstack.top();
        }
    }

    int getMin() {
        if (!minstack.empty()) {
            return minstack.top();
        }
    }

    stack<int> vstack;
    stack<int> minstack;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: leetcode-algorithms/155. Min Stac/Solution.java
class MinStack {

    public MinStack() {

    }

    public void push(int x) {
        if (!minstack.isEmpty()) {
            minstack.add(Math.min(minstack.peek(), x));
            vstack.add(x);
        } else {
            vstack.add(x);
            minstack.add(x);
        }
    }

    public void pop() {
        if (!vstack.isEmpty()) {
            vstack.pop();
            minstack.pop();
        }
    }

    public int top() {
        if (vstack.isEmpty()) return -1;

        return vstack.peek();
    }

    public int getMin() {
        if (vstack.isEmpty()) return -1;

        return minstack.peek();
    }

    Stack<Integer> vstack = new Stack<>();
    Stack<Integer> minstack = new Stack<>();
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
